File: tests/preview5.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {createPathSampler} from '../src/core/path-sampler.js';
import {createRainSystem} from '../src/core/rain-system.js';
import {seaSurface} from '../src/core/sea-model.js';
import {createMetaballMotion} from '../src/core/metaball-motion.js';
import {lightningGeometry} from '../src/core/lightning-geometry.js';
import {allowsPointer} from '../src/core/interaction.js';
import {getDefaults,normalizeOptions} from '../src/core/schema.js';
import {resolveMotion,presentationOptions} from '../src/core/motion.js';
const near=(a,b,epsilon=1e-7)=>assert.ok(Math.abs(a-b)<=epsilon,`${a} != ${b}`);
const options=(family,patch={})=>normalizeOptions(family,patch,getDefaults(family));
const sample=(points,config={})=>{const out=[],s=createPathSampler(config);for(const p of points)s.push(p,x=>out.push(x));return out;};
const coords=points=>points.map(p=>[p.x,p.y]);
test('Distance throttle emits equal CSS spacing, independently of pointer event rate',()=>{
 const a=sample([{x:0,y:0,time:0},{x:240,y:0,time:1000}]);
 const b=sample(Array.from({length:241},(_,i)=>({x:i,y:0,time:i*1000/240})));
 assert.deepEqual(coords(a),coords(b));assert.equal(a.length,20);assert.deepEqual(coords(a),Array.from({length:20},(_,i)=>[(i+1)*12,0]));
});
test('Distance remainder follows bends instead of cutting corner-to-corner',()=>{
 const out=sample([{x:0,y:0,time:0},{x:8,y:0,time:10},{x:8,y:16,time:20},{x:12,y:16,time:30}],{distance:10});
 assert.deepEqual(coords(out),[[8,2],[8,12]]);
});
test('A parked pointer cannot emit distance-throttled particles',()=>{
 assert.equal(sample([{x:0,y:0,time:0},{x:0,y:0,time:1e9}]).length,0);
});
test('Time throttle interpolates samples at the configured interval',()=>{
 const a=sample([{x:0,y:0,time:0},{x:100,y:50,time:100}],{mode:'time',interval:20});
 assert.deepEqual(coords(a),[[20,10],[40,20],[60,30],[80,40],[100,50]]);
 const b=sample(Array.from({length:11},(_,i)=>({x:i*10,y:i*5,time:i*10})),{mode:'time',interval:20});assert.deepEqual(coords(a),coords(b));
});
test('Time mode supports stationary hover, but limits extreme catch-up bursts',()=>{
 const out=sample([{x:20,y:20,time:0},{x:20,y:20,time:100000}],{mode:'time',interval:10,limit:20});assert.equal(out.length,20);assert.ok(out[0].time>99000);
});
test('Exit/reset, malformed coordinates and clock rewind never bridge strokes',()=>{
 const s=createPathSampler(),out=[],emit=x=>out.push(x);s.push({x:0,y:0,time:0},emit);s.reset();s.push({x:1000,y:0,time:10},emit);assert.equal(out.length,0);
 s.push({x:NaN,y:0,time:11},emit);s.push({x:0,y:0,time:12},emit);assert.equal(out.length,0);
 s.push({x:0,y:0,time:0},emit);s.push({x:12,y:0,time:1},emit);assert.equal(out.length,1);
});
test('Sampler work stays bounded and partial configuration retains interval',()=>{
 const s=createPathSampler({distance:5,interval:40,limit:24});s.push({x:0,y:0,time:0},()=>{});let n=0;s.push({x:1000000,y:0,time:100},()=>n++);assert.equal(n,24);assert.ok(s.getState().dropped>1000);
 s.configure({mode:'time'});assert.equal(s.getState().interval,40);assert.equal(s.getState().distance,5);
});
test('Default trail throttle is distance; interval aliases and bounds are normalized',()=>{
 const o=options('trail');assert.equal(o.throttle,'distance');assert.equal(o.trigger,'move');assert.ok(o.distanceInterval>0);
 near(normalizeOptions('trail',{emission:100},o).timeInterval,10);
 near(normalizeOptions('trail',{timeInterval:50},o).emission,20);
 assert.equal(options('trail',{distanceInterval:0}).distanceInterval,1);assert.equal(options('trail',{timeInterval:0}).timeInterval,4);
});
test('Rain integrates velocity in CSS pixels per second without random per-frame directions',()=>{
 const o=options('field',{kind:'rain',rainSpeed:600,rainWind:45,rainGust:0,rainDepth:.5}),r=createRainSystem(o);r.resize(600,400);r.step(0);
 const p=r.drops.find(p=>p.y>0&&p.y<250&&p.x>50&&p.x<500),x=p.x,y=p.y;r.step(.05);near(p.x-x,p.vx*.05);near(p.y-y,p.vy*.05);assert.ok(p.vy>0);
 const oldX=p.x,oldY=p.y;r.setOptions({...o,rainSpeed:1200});near(p.x,oldX);near(p.y,oldY);r.step(.10);near(p.y-oldY,1200*p.speed*.05);
});
test('Rain frame integration is stable at different frame rates in constant wind',()=>{
 const o=options('field',{kind:'rain',rainSpeed:100,rainWind:12,rainGust:0}),a=createRainSystem(o),b=createRainSystem(o);for(const r of[a,b]){r.resize(900,900);r.step(0);}
 for(let i=1;i<=60;i++)a.step(i/60);for(let i=1;i<=30;i++)b.step(i/30);
 const i=a.drops.findIndex((p,i)=>p.y>200&&p.y<600&&b.drops[i].y>200);near(a.drops[i].x,b.drops[i].x);near(a.drops[i].y,b.drops[i].y);
});
test('Rain keeps CSS speed and gust trajectories even at one frame per second',()=>{
 const o=options('field',{kind:'rain',rainSpeed:1200,rainWind:150,rainGust:.6}),a=createRainSystem(o),b=createRainSystem(o);for(const r of[a,b]){r.resize(600,300);r.step(0);}
 a.step(1);for(let i=1;i<=120;i++)b.step(i/120);
 for(let i=0;i<a.drops.length;i++){near(a.drops[i].x,b.drops[i].x,1e-6);near(a.drops[i].y,b.drops[i].y,1e-6);}
 assert.equal(a.getState().recycled,b.getState().recycled);
});
test('Rain pool is bounded, seeded and can be completely empty',()=>{
 const o=options('field',{kind:'rain',rainDensity:3}),a=createRainSystem(o),b=createRainSystem(o);a.resize(10000,9000);b.resize(10000,9000);assert.equal(a.drops.length,1100);assert.deepEqual(a.drops,b.drops);
 a.setQuality(.2);assert.ok(a.drops.length<=1100);a.setOptions({...o,rainDensity:0});assert.equal(a.drops.length,0);
});
test('Sea height and both slopes become exactly flat at zero amplitude',()=>{
 const o=options('sea',{waveHeight:0});for(let t=0;t<4;t+=.1)assert.deepEqual(seaSurface(3,8,t,o),{height:0,dx:0,dz:0});
});
test('Sea at a fixed world location changes between crest and trough',()=>{
 const o=options('sea');const heights=Array.from({length:600},(_,i)=>seaSurface(1,3,i/30,o).height);
 assert.ok(Math.min(...heights)<-.04);assert.ok(Math.max(...heights)>.04);assert.ok(Math.max(...heights)-Math.min(...heights)>.14);
});
test('Sea analytical slopes match numerical derivatives of the signed height field',()=>{
 const o=options('sea'),eps=1e-5;for(let i=0;i<20;i++){const x=i*.17,z=i*.27,t=i*.3,a=seaSurface(x,z,t,o,20);
 near(a.dx,(seaSurface(x+eps,z,t,o,20).height-seaSurface(x-eps,z,t,o,20).height)/(2*eps),1e-6);
 near(a.dz,(seaSurface(x,z+eps,t,o,20).height-seaSurface(x,z-eps,t,o,20).height)/(2*eps),1e-6);}
});
test('Sea direction rotates the wave field, not the camera; zero phase speed is static',()=>{
 const o=options('sea',{waveDirection:0}),a=seaSurface(2,5,3,o),b=seaSurface(-5,2,3,{...o,waveDirection:90});near(a.height,b.height);
 assert.deepEqual(seaSurface(2,5,1,{...o,waveSpeed:0}),seaSurface(2,5,5,{...o,waveSpeed:0}));
});
test('Pointer-only metal is at rest until a local brush acts on multiple nearby bodies',()=>{
 const o=options('art',{kind:'metaballs',interactionMode:'pointer'}),m=createMetaballMotion(o);m.resize(600,300);
 const before=Array.from(m.uniforms()['uBodies[0]']);for(let i=0;i<30;i++)m.step(i/60);assert.deepEqual(Array.from(m.uniforms()['uBodies[0]']),before);assert.equal(m.animated,false);
 m.setPointer({x:.2,y:.5,active:true});m.setPointer({x:.8,y:.5,active:true});assert.ok(m.getState().affectedBodies>=2);assert.ok(m.animated);
 for(let i=30;i<45;i++)m.step(i/60);assert.notDeepEqual(Array.from(m.uniforms()['uBodies[0]']),before);
 for(let i=45;i<3600;i++)m.step(i/60);assert.ok(m.getState().kineticEnergy<.00001);assert.equal(m.animated,false);
});
test('Metal Auto ignores pointer; Press requires a held contact; exit does not kick it',()=>{
 const o=options('art',{kind:'metaballs',interactionMode:'auto'}),m=createMetaballMotion(o);m.resize(600,300);
 m.setPointer({x:.2,y:.5,active:true});m.setPointer({x:.8,y:.5,active:true});assert.equal(m.getState().affectedBodies,0);
 m.setOptions({...o,interactionMode:'pointer',interactionTrigger:'press'});m.setPointer({x:.2,y:.5,active:true,down:false});m.setPointer({x:.8,y:.5,active:true,down:false});assert.equal(m.animated,false);
 m.setPointer({x:.2,y:.5,active:true,down:true});m.setPointer({x:.8,y:.5,active:true,down:true});assert.ok(m.getState().affectedBodies>=2);
 m.reset();m.suspend();m.setPointer({x:.9,y:.9,dx:1,dy:1,active:true,down:true});assert.equal(m.animated,false);
});
test('Metal simulation stays finite and bounded under repeated strong strokes',()=>{
 const o=options('art',{kind:'metaballs',interactionStrength:3,viscosity:1,interactionMode:'both'}),m=createMetaballMotion(o);m.resize(600,300);
 for(let i=0;i<1200;i++){m.setPointer({active:true,x:i%2?.2:.8,y:.5});m.step(i/60);for(const v of m.uniforms()['uBodies[0]'])assert.ok(Number.isFinite(v)&&Math.abs(v)<=1.600001);}
});
test('Lightning is reproducible, multiscale, branched and geometrically bounded',()=>{
 const o={seed:11,start:{x:300,y:10},end:{x:410,y:290},width:600,height:300,branches:7};const a=lightningGeometry(o),b=lightningGeometry(o);assert.deepEqual(a,b);assert.ok(a.length>=8&&a.length<=22);assert.ok(a[0].path.length>=65);assert.deepEqual(a[0].path[0],o.start);assert.deepEqual(a[0].path.at(-1),o.end);
 for(const path of a)for(const p of path.path)assert.ok(Number.isFinite(p.x)&&Number.isFinite(p.y)&&Math.abs(p.x)<1500&&Math.abs(p.y)<1000);
 assert.notDeepEqual(a,lightningGeometry({...o,seed:12}));assert.ok(a.reduce((sum,p)=>sum+p.path.length,0)<2500);
});
test('Lightning branches originate on the parent leader rather than unrelated straight spokes',()=>{
 const a=lightningGeometry({start:{x:300,y:0},end:{x:320,y:280},branches:8});for(const b of a.filter(b=>b.depth===1))assert.ok(a[0].path.some(p=>p===b.path[0]));
 assert.throws(()=>lightningGeometry({start:{x:NaN,y:0},end:{x:3,y:4}}),TypeError);
});
test('Only meaningful interactions accept pointer input, with explicit source choices',()=>{
 for(const effect of['water','sea','glow','fire','aurora','clouds'])assert.equal(allowsPointer(effect,options(effect)),false);
 for(const kind of['silk','tunnel'])assert.equal(allowsPointer('art',options('art',{kind})),false);
 for(const kind of['orbit','jellyfish'])assert.equal(allowsPointer('sketch',options('sketch',{kind})),false);
 assert.equal(allowsPointer('art',options('art',{kind:'metaballs',interactionMode:'both'})),true);
 assert.equal(allowsPointer('art',options('art',{kind:'metaballs',interactionMode:'auto'})),false);
 for(const trigger of['click','both'])assert.equal(allowsPointer('sketch',options('sketch',{kind:'lightning',strikeTrigger:trigger})),true);
});
test('Reduced motion adapts the sea and suppresses transients without mutating saved options',()=>{
 const sea=options('sea'),calm=presentationOptions('sea',sea,resolveMotion('sea',sea,true));assert.equal(calm.waveHeight,0);assert.equal(calm.ripples,0);assert.ok(sea.waveHeight>0);
 const rain=options('field',{kind:'rain'});assert.equal(resolveMotion('field',rain,true),'hide');
 for(const reducedMotion of['auto','calm','subtle'])assert.equal(resolveMotion('sketch',options('sketch',{kind:'lightning',reducedMotion}),true),'hide');
 assert.equal(resolveMotion('sea',options('sea',{motion:'always'}),true),'full');assert.equal(resolveMotion('sea',options('sea',{motion:'never'}),false),'calm');
});

File: tests/preview5_test.py
"""Preview 5: real Canvas pixels, pointer lifecycle and live custom-element controls.

WebGL execution is tested separately through EGL, not imitated by these checks.
"""
import json
import mimetypes
from urllib.parse import urlparse
import os
import shutil
from pathlib import Path
from playwright.sync_api import sync_playwright

ROOT=Path(__file__).resolve().parents[1]
OUT=ROOT/'tests/validation'
OUT.mkdir(exist_ok=True)
checks=[]
def check(name,value):
    checks.append({'name':name,'passed':bool(value)})
    assert value,name
def main():
    with sync_playwright() as p:
        browser=p.chromium.launch(headless=True,executable_path=os.environ.get('CHROMIUM_PATH') or shutil.which('chromium'))
        page=browser.new_page(viewport={'width':1100,'height':850},reduced_motion='no-preference')
        errors=[];page.on('pageerror',lambda error:errors.append(str(error)))
        page.set_content('<div id="mount" style="position:relative;width:600px;height:300px;border-radius:24px"></div>')
        page.add_script_tag(content=(ROOT/'dist/frontend-toolkit.js').read_text())
        results=page.evaluate('''()=>{
          const FT=FrontendToolkit,tests=[],check=(name,ok)=>{tests.push({name,passed:!!ok});if(!ok)throw Error(name);};
          const config=(f,p={})=>({...FT.getDefaults(f),adaptive:false,dprCap:1,...p});
          const hash=c=>{const d=c.getContext('2d').getImageData(0,0,c.width,c.height).data;let h=2166136261,a=0;for(let i=0;i<d.length;i++){h=Math.imul(h^d[i],16777619);if(i%4===3)a+=d[i];}return {h:h>>>0,a};};
          function brush(patch={}){const r=FT.createTrailRenderer(mount,config('trail',{life:5,count:800,distanceInterval:20,size:4,spread:0,turbulence:0,gravity:0,...patch}));r.resize(600,300);r.render(0);return r;}
          function stroke(r,n){for(let i=0;i<=n;i++)r.setPointer({x:.1+.8*i/n,y:.5,active:true,timeStamp:i*100/n,event:i?'pointermove':'pointerenter'});r.render(.10);}
          let r=brush();stroke(r,1);check('Coarse pointer stroke emits 24 distance-spaced particles',r.getState().spawned===24);check('Distance trail paints actual native Canvas pixels',hash(r.canvas).a>0);const a=r.getState().spawned;r.render(.2);check('Stationary move-only brush emits no more particles',r.getState().spawned===a);r.destroy();
          r=brush();stroke(r,48);check('Dense pointer events produce the same emission count',r.getState().spawned===24);r.destroy();
          r=brush({throttle:'time',timeInterval:20,trigger:'hover'});r.setPointer({x:.5,y:.5,active:true,timeStamp:0});for(let i=1;i<=5;i++)r.render(i*.02);check('Time interval controls stationary hover emission',r.getState().spawned===5);r.suspendInput();const n=r.getState().spawned;r.render(.2);check('Input suspension drains no stale hover samples',r.getState().spawned===n);r.destroy();
          r=brush({trigger:'press'});r.setPointer({x:.2,y:.5,active:true,down:true,timeStamp:0});r.render(.1);check('Distance press waits for actual movement',r.getState().spawned===0);r.setPointer({x:.8,y:.5,active:true,down:true,timeStamp:100});r.render(.2);check('Held distance brush emits along its stroke',r.getState().spawned===18);r.setPointer({x:2,y:.5,active:false,event:'pointercancel'});r.setPointer({x:.1,y:.5,active:true,down:true,timeStamp:200});r.render(.3);check('Reentry after cancellation does not connect distant strokes',r.getState().spawned===18);r.destroy();
          r=brush({trigger:'both'});stroke(r,1);for(let i=1;i<=30;i++)r.render(.1+i/60);check('Both combines autonomous and pointer emissions',r.getState().spawned>24);r.destroy();
          r=brush();r.setPointer({x:0,y:.5,active:true,timeStamp:0});r.setPointer({x:1,y:.5,active:true,timeStamp:10});r.setOptions({distanceInterval:8});r.render(.1);check('Changing throttle resets queued old samples',r.getState().spawned===0);r.destroy();
          r=brush({trigger:'move',throttle:'time',timeInterval:20});r.setPointer({x:.2,y:.5,active:true,timeStamp:0});r.setPointer({x:.3,y:.5,active:true,timeStamp:60});r.render(.1);const early=r.getState().spawned;r.setPointer({x:.4,y:.5,active:true,timeStamp:3000});r.render(.2);check('Time-based move trigger does not replay seconds of idle cursor history',r.getState().spawned===early+1);r.destroy();
          let contexts=0;const get=HTMLCanvasElement.prototype.getContext;HTMLCanvasElement.prototype.getContext=function(type,...a){if(/webgl/.test(type))contexts++;return get.call(this,type,...a);};
          r=FT.createFieldRenderer(mount,config('field',{kind:'rain',transparent:true,rainGust:0,rainWind:40}));r.resize(600,300);r.render(0);const rain0=hash(r.canvas);r.render(.05);const rain1=hash(r.canvas);check('Rain animates native pixels with no WebGL allocation',rain0.h!==rain1.h&&rain1.a>0&&contexts===0);check('Rain diagnostics expose explicit CSS-pixel speed',r.getState().rainSpeed===620&&r.getState().rainWind===40);
          r.setOptions({rainSpeed:0,rainWind:0,rainGust:0});r.render(.1);const still=hash(r.canvas);r.render(.2);check('Zero rain velocity is static, without random jitter',hash(r.canvas).h===still.h&&!r.animated);
          r.setOptions({rainDensity:0});r.render(.3);check('Zero rain density leaves a transparent empty layer',hash(r.canvas).a===0&&r.getState().drops===0);
          r.setOptions({rainDensity:1,rainSpeed:800});r.resize(10000,9000);r.render(.4);check('Rain pixel and particle budgets are bounded on huge hosts',r.canvas.width*r.canvas.height<=1250000&&Math.max(r.canvas.width,r.canvas.height)<=2048&&r.getState().drops<=1100);
          r.setOptions({kind:'ocean',forceFallback:true});r.resize(600,300);r.render(.5);check('Switching Rain to a GPU-family fallback replaces the backend cleanly',r.getState().backend==='Canvas 2D fallback'&&mount.querySelectorAll('canvas').length===1);
          r.setOptions({kind:'rain'});r.render(.6);check('Switching back reuses the field API without a duplicate canvas',r.getState().backend==='Canvas layered rain'&&mount.querySelectorAll('canvas').length===1);r.destroy();HTMLCanvasElement.prototype.getContext=get;
          r=FT.createArtRenderer(mount,config('art',{forceFallback:true,kind:'metaballs',interactionMode:'pointer'}));r.resize(600,300);r.render(0);const before=hash(r.canvas);r.render(.1);check('Pointer-only metal rests without autonomous drift',before.h===hash(r.canvas).h&&!r.animated);
          r.setPointer({x:.2,y:.5,active:true});r.setPointer({x:.8,y:.5,active:true});for(let i=1;i<=20;i++)r.render(.1+i/60);check('Local metal brush reaches several nearby surfaces',r.getState().affectedBodies>=2);check('Local metal physics changes actual fallback pixels',before.h!==hash(r.canvas).h);
          r.setOptions({interactionMode:'auto'});const energy=r.getState().kineticEnergy;r.setPointer({x:.1,y:.1,active:true});r.setPointer({x:.9,y:.9,active:true});check('Auto mode does not inject pointer impulses',r.getState().kineticEnergy===energy);r.destroy();
          for(const [family,kind] of [['art','silk'],['art','tunnel'],['sketch','orbit'],['sketch','jellyfish'],['field','lava'],['field','nebula'],['field','ocean'],['sea',null]]){
            const renderer=FT['create'+family[0].toUpperCase()+family.slice(1)+'Renderer'](mount,config(family,{forceFallback:true,...(kind?{kind}:{})}));renderer.resize(480,260);renderer.render(.2);const first=hash(renderer.canvas);renderer.setPointer?.({x:.1,y:.1,active:true,dx:-.4,dy:-.4});renderer.render(.2);renderer.setPointer?.({x:.9,y:.9,active:true,dx:.8,dy:.8});renderer.render(.2);check((kind||family)+': no opposite-pointer scene/camera motion',hash(renderer.canvas).h===first.h);renderer.destroy();
          }
          for(const [family,kind] of [['art','interference'],['field','magnetic'],['sketch','constellation'],['sketch','blobs'],['smoke',null]]){
            const renderer=FT['create'+family[0].toUpperCase()+family.slice(1)+'Renderer'](mount,config(family,{forceFallback:true,interactionMode:'pointer',...(kind?{kind}:{})}));renderer.resize(480,260);renderer.render(.2);renderer.setPointer({x:.2,y:.3,active:true,dx:0,dy:0});renderer.render(.2);const first=hash(renderer.canvas);renderer.setPointer({x:.7,y:.6,active:true,dx:.5,dy:.3});renderer.render(.2);check((kind||family)+': meaningful local interaction remains available',hash(renderer.canvas).h!==first.h);check((kind||family)+': pointer-only mode needs no continuous frame job',!renderer.animated);renderer.destroy();
          }
          r=FT.createSketchRenderer(mount,config('sketch',{kind:'lightning',strikeTrigger:'manual'}));r.resize(600,300);r.render(0);check('Manual lightning is absent until explicitly triggered',hash(r.canvas).a===0&&!r.animated);r.burst({x:.75,y:.1});r.render(.02);r.render(.08);check('Lightning has irregular multiscale channels, not sparse spokes',r.getState().lightningVertices>100&&r.getState().lightningChannels>1);check('A discharge paints a narrow luminous channel',hash(r.canvas).a>0);r.render(2);check('Manual lightning settles to transparency and releases its clock',hash(r.canvas).a===0&&!r.animated);r.destroy();
          check('All renderer teardown paths remove their canvas nodes',mount.childElementCount===0);return tests;
        }''')
        checks.extend(results)
        # Actual custom elements: no direct renderer shortcuts for input/motion tests.
        page.evaluate("mount.innerHTML='<ft-trail id=fx throttle=time time-interval=15 trigger=hover><button id=child>Real button</button></ft-trail>';fx.style.cssText='width:600px;height:300px';")
        page.wait_for_function('fx.controller?.renderer')
        page.dispatch_event('#fx','pointerenter',{'clientX':120,'clientY':120,'buttons':0,'pointerType':'mouse','isPrimary':True})
        page.wait_for_timeout(140)
        check('Declarative time throttle accepts a stationary hover',page.evaluate('fx.getStats().spawned>0'))
        page.evaluate('fx.pause()');page.wait_for_timeout(30)
        paused=page.evaluate('fx.getStats().spawned')
        page.dispatch_event('#fx','pointermove',{'clientX':530,'clientY':150,'buttons':0,'pointerType':'mouse','isPrimary':True})
        page.evaluate('fx.play()');page.wait_for_timeout(140)
        check('Pause/resume does not replay an old pointer sample or held emitter',page.evaluate('fx.getStats().spawned')==paused)
        page.dispatch_event('#fx','pointerenter',{'clientX':130,'clientY':150,'buttons':0,'pointerType':'mouse','isPrimary':True});page.wait_for_timeout(80)
        page.dispatch_event('#fx','pointercancel',{'clientX':130,'clientY':150,'buttons':0,'pointerType':'mouse','isPrimary':True});page.wait_for_timeout(20)
        cancelled=page.evaluate('fx.getStats().spawned');page.wait_for_timeout(120)
        check('Pointer cancellation ends stationary emission',page.evaluate('fx.getStats().spawned')==cancelled)
        page.dispatch_event('#child','pointermove',{'clientX':15,'clientY':15,'buttons':0,'pointerType':'mouse','isPrimary':True});page.wait_for_timeout(100)
        check('Native content controls do not start a decorative brush',page.evaluate('fx.getStats().spawned')==cancelled)
        page.evaluate("fx.configure({trigger:'move',throttle:'distance',distanceInterval:20});fx.clear()")
        page.wait_for_timeout(60)
        emitted=page.evaluate('fx.getStats().spawned')
        page.evaluate('''()=>{
          const b=fx.getBoundingClientRect(),y=b.top+150;
          fx.dispatchEvent(new PointerEvent('pointerenter',{clientX:b.left+60,clientY:y,isPrimary:true}));
          const e=new PointerEvent('pointermove',{clientX:b.left+540,clientY:y,isPrimary:true});
          Object.defineProperty(e,'getCoalescedEvents',{value:()=>[new PointerEvent('pointermove',{clientX:b.left+300,clientY:y}),new PointerEvent('pointermove',{clientX:b.left+540,clientY:y})]});fx.dispatchEvent(e);
        }''');page.wait_for_timeout(80)
        check('Optional coalesced input preserves CSS-distance spacing',page.evaluate('fx.getStats().spawned')-emitted==24)
        page.emulate_media(reduced_motion='reduce');page.wait_for_timeout(80)
        check('Reduced motion suppresses trails and discards the input queue',page.evaluate("fx.getStats().motionPresentation==='hide'&&fx.controller.renderer===null"))
        page.evaluate("mount.innerHTML='<ft-sea id=fx></ft-sea>';fx.style.cssText='width:600px;height:300px';fx.configure({forceFallback:true})")
        page.wait_for_function('fx.controller?.renderer')
        check('Reduced-motion sea uses flat geometry, not a frozen crest',page.evaluate("fx.getStats().motionPresentation==='calm'&&fx.controller.renderer.getUniforms().uAmplitude===0&&fx.options.waveHeight>0"))
        t=page.evaluate('fx.getStats().time');page.wait_for_timeout(120)
        check('Calm sea requests no animation frames',page.evaluate('fx.getStats().time')==t and page.evaluate('!fx.getStats().scheduled'))
        page.evaluate("fx.configure({motion:'always'})");page.wait_for_timeout(220)
        check('Explicit motion override restores waves and managed time',page.evaluate('fx.controller.renderer.getUniforms().uAmplitude>0&&fx.getStats().time>0'))
        page.evaluate("fx.configure({motion:'never',reducedMotion:'subtle'})");page.wait_for_timeout(80)
        check('Never cannot accidentally enable subtle autonomous waves',page.evaluate("!fx.getStats().animating&&fx.getStats().motionPresentation==='calm'"))
        page.evaluate("mount.innerHTML='<ft-sketch id=fx kind=lightning reduced-motion=subtle></ft-sketch>';fx.style.cssText='width:600px;height:300px'");page.wait_for_timeout(100)
        check('Reduced-motion lightning is suppressed even in subtle mode',page.evaluate("fx.getStats().motionPresentation==='hide'"))
        page.emulate_media(reduced_motion='no-preference');page.evaluate('mount.replaceChildren()');page.wait_for_timeout(150)
        check('Controller listeners and frame jobs are cleaned up',page.evaluate('FrontendToolkit.schedulerStats().jobs===0'))
        # Screenshots use actual native Canvas renderers.
        for family,preset in [('sketch','lightning'),('field','rain')]:
            page.evaluate('''([family,preset])=>{mount.replaceChildren();mount.style.background='#071017';const r=FrontendToolkit['create'+family[0].toUpperCase()+family.slice(1)+'Renderer'](mount,{...FrontendToolkit.getDefaults(family,preset),quality:1,dprCap:1});r.resize(600,300);r.render(0);r.render(.07);window.previewRenderer=r;}''',[family,preset])
            page.locator('#mount').screenshot(path=str(OUT/f'preview5-{preset}.png'))
            page.evaluate('previewRenderer.destroy()')
        # The environment blocks top-level navigation. Serve unchanged resource
        # bytes through an offline HTTPS fixture; this is not a network test.
        page.close();page=browser.new_page(viewport={'width':1100,'height':850},reduced_motion='no-preference');page.on('pageerror',lambda error:errors.append(str(error)))
        base='https://toolkit.test'
        def serve(route):
            path=(ROOT/urlparse(route.request.url).path.lstrip('/')).resolve()
            if path.is_file() and path.is_relative_to(ROOT):
                route.fulfill(status=200,body=path.read_bytes(),headers={'Content-Type':mimetypes.guess_type(path)[0] or 'application/octet-stream','Access-Control-Allow-Origin':'*'})
            else:route.fulfill(status=404,body='missing')
        page.route(base+'/**',serve)
        page.set_content((ROOT/'lab.html').read_text().replace('<head>','<head><base href="https://toolkit.test/">'))
        page.wait_for_function('window.MaterialLab&&window.ToolkitLab')
        page.locator('#effect-nav [data-id="rain"]').click();page.wait_for_timeout(150)
        check('Rain speed has a visible dedicated inspector control',page.locator('#option-rainSpeed').is_visible())
        check('Rain exposes its actual droplet color rather than an unused shader parameter',page.locator('#option-color2').is_visible())
        page.locator('#effect-nav [data-id="stars"]').click();page.wait_for_timeout(80)
        check('Trail inspector defaults to distance spacing',page.locator('#option-throttle').input_value()=='distance' and page.locator('#option-distanceInterval').is_visible())
        page.select_option('#option-throttle','time');page.wait_for_timeout(80)
        check('Choosing Time reveals milliseconds and hides pixel spacing',page.locator('#option-timeInterval').is_visible() and page.locator('#option-distanceInterval').count()==0)
        page.locator('#effect-nav [data-id="metaballs"]').click();page.wait_for_timeout(80)
        check('Liquid metal exposes motion source and local brush controls',page.locator('#option-interactionMode').is_visible() and page.locator('#option-interactionRadius').is_visible())
        page.locator('#effect-nav [data-id="silk"]').click();page.wait_for_timeout(60)
        check('Pure compositions do not expose misleading pointer-camera controls',page.locator('#option-interactive').count()==0 and page.locator('#option-interactionMode').count()==0)
        page.locator('#effect-nav [data-id="lightning"]').click();page.wait_for_timeout(60)
        check('Lightning has distinct strike-trigger and irregularity controls',page.locator('#option-strikeTrigger').is_visible() and page.locator('#option-jaggedness').is_visible())
        page.screenshot(path=str(OUT/'preview5-demo-desktop.png'))
        page.set_viewport_size({'width':390,'height':844});page.wait_for_timeout(100)
        check('Narrow demo layout does not create horizontal page overflow',page.evaluate('document.documentElement.scrollWidth<=innerWidth+1'))
        page.screenshot(path=str(OUT/'preview5-demo-mobile.png'))
        check('No uncaught exceptions in native renderers or demo',not errors)
        selective=browser.new_page(viewport={'width':900,'height':600});fetched=[]
        selective.on('request',lambda request:fetched.append(request.url))
        selective.route(base+'/**',serve)
        selective.set_content('<ft-trail id=fx trigger=manual style="width:500px;height:220px"></ft-trail>')
        selective.add_script_tag(type='module',content="import * as FT from 'https://toolkit.test/dist/frontend-toolkit.loader.js';window.fixtureToolkit=FT;")
        selective.wait_for_function('document.querySelector("#fx").controller?.renderer')
        check('Selective and classic entries report the same current version',selective.evaluate("fixtureToolkit.version==='1.0.0'"))
        check('Selective trail loading fetches the new sampler transitively',any('/core/path-sampler.js' in url for url in fetched))
        check('Selective trail loading does not fetch sea or metal renderers',not any('/shaders/sea.js' in url or '/shaders/art.js' in url for url in fetched))
        selective.evaluate("fx.remove();const m=document.createElement('ft-art');m.style.cssText='width:500px;height:220px';m.id='metal';m.setAttribute('force-fallback','true');document.body.append(m)")
        selective.wait_for_function('document.querySelector("#metal").controller?.renderer')
        check('Selective metal loading includes its local physics module',any('/core/metaball-motion.js' in url for url in fetched))
        browser.close()
    (OUT/'preview5-browser-report.json').write_text(json.dumps({'checks':checks,'nativeCanvasChecks':len(checks),'nativeWebGLExecution':False},indent=2))
    print(f'{len(checks)} Preview 5 native-browser checks passed (Canvas, interactions, motion policy, demo and selective imports).')
if __name__=='__main__':main()
